File: tests/cs-sdk/TestRequestResponseIntegrationLoop.cs
// loop test: Request-Response (Querier + Responser)
using System;
using System.Threading;
using System.Threading.Tasks;
using KnotLink;

const int MAX_ITERATIONS = 5;
var cts = new CancellationTokenSource();

var t = Task.Run(() => RunResponser(cts.Token));
await Task.Delay(100);
await RunQuerier(cts.Token);
cts.Cancel();
await t;

static async Task RunResponser(CancellationToken ct)
{
    var res = new OpenSocketResponser("app.knotlinksdk.test", "test_socket")
    {
        OnQuestionAsync = data =>
        {
            Console.WriteLine($"[Responser] Received: {data}");
            return Task.FromResult($"Echo: {data}");
        }
    };
    Console.WriteLine("[Responser] Running...");
    while (!ct.IsCancellationRequested)
        await Task.Delay(1000, ct);
}

static async Task RunQuerier(CancellationToken ct)
{
    await Task.Delay(1000, ct);
    var q = new OpenSocketQuerier("app.knotlinksdk.test", "test_socket");
    try
    {
        for (int i = 1; i <= MAX_ITERATIONS && !ct.IsCancellationRequested; i++)
        {
            var data = $"Hello #{i}";
            try
            {
                var result = q.Query(data);
                Console.WriteLine($"[Querier] Req: {data} -> Res: {result}");
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"[Querier] Error: {ex.Message}");
            }
            await Task.Delay(3000, ct);
        }
    }
    finally { q.Dispose(); }
}

File: knotlink-sdk-cpp/tests/test_request_response_integration_loop.cpp
// loop test: Request-Response (Querier + Responser)
#include "../knotlink/OpenSocketResponser.hpp"
#include "../knotlink/OpenSocketQuerier.hpp"
#include <iostream>
#include <thread>
#include <chrono>
#include <atomic>
using namespace knotlink;

static std::atomic<bool> shutdown{false};
constexpr int MAX_ITERATIONS = 5;

void run_responser() {
    OpenSocketResponser res("app.knotlinksdk.test", "test_socket");
    res.setQuestionHandler([](const std::string& data) -> std::string {
        std::cout << "[Responser] Received: " << data << std::endl;
        return "Echo: " + data;
    });
    std::cout << "[Responser] Running..." << std::endl;
    while (!shutdown) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
}

void run_querier() {
    std::this_thread::sleep_for(std::chrono::seconds(1));
    OpenSocketQuerier q;
    q.setConfig("app.knotlinksdk.test", "test_socket");
    for (int i = 1; i <= MAX_ITERATIONS && !shutdown; i++) {
        try {
            std::string data = "Hello #" + std::to_string(i);
            std::string result = q.query_l(data);
            std::cout << "[Querier] Req: " << data << " -> Res: " << result << std::endl;
        } catch (const std::exception& e) {
            std::cerr << "[Querier] Error: " << e.what() << std::endl;
        }
        std::this_thread::sleep_for(std::chrono::seconds(3));
    }
    shutdown = true;
}

int main() {
    std::thread t(run_responser);
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    run_querier();
    t.join();
    return 0;
}
